src/control/security/pem.go

```diff
@@ -0,0 +1,107 @@
+package security
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+)
+
+const (
+	SafeKeyPerm  os.FileMode = 0400
+	SafeCertPerm os.FileMode = 0664
+	SafeDirPerm  os.FileMode = 0700
+)
+
+type insecureError struct {
+	filename string
+	perms    os.FileMode
+}
+
+func (e *insecureError) Error() string {
+	return fmt.Sprintf("%s has insecure permissions set to %#o",
+		e.filename, e.perms)
+}
+
+func LoadPEMData(filePath string, perms os.FileMode) ([]byte, error) {
+	statInfo, err := os.Stat(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	if !statInfo.Mode().IsRegular() {
+		return nil, fmt.Errorf("%s is not a regular file", filePath)
+	}
+	if statInfo.Mode().Perm() != perms {
+		return nil, &insecureError{filePath, perms}
+	}
+
+	fileContents, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+	return fileContents, nil
+}
+
+func LoadPrivateKey(keyPath string) (crypto.PrivateKey, error) {
+
+	pemData, err := LoadPEMData(keyPath, SafeKeyPerm)
+
+	if err != nil {
+		return nil, err
+	}
+
+	block, extra := pem.Decode(pemData)
+
+	if block == nil {
+		return nil, fmt.Errorf("%s does not contain PEM data", keyPath)
+	}
+
+	if len(extra) != 0 {
+		return nil, fmt.Errorf("Only one key allowed per file")
+	}
+
+	/* covers multiple key types */
+	if !strings.HasSuffix(block.Type, "PRIVATE KEY") {
+		return nil, fmt.Errorf("PEM Block is not a Private Key")
+	}
+
+	// Based on parsePrivateKey in golang.org/src/crypto/tls/tls.go
+	// Pre OpenSSL 1.0.0 default key packaging
+	rsaKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err == nil {
+		return rsaKey, nil
+	}
+
+	// Post OpenSSL 1.0.0 default key packaging
+	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if err == nil {
+		switch key := key.(type) {
+		// TODO: Support key types other than RSA
+		case *rsa.PrivateKey:
+			return key, nil
+		default:
+			return nil, fmt.Errorf("%s contains an unsupported private key type",
+				keyPath)
+		}
+	}
+	return key, nil
```

Comment:
Shouldn't this default to returning an error?  e.g.:
```suggestion
	return nil, fmt.Errorf("Unhandled error: %s", err)
```

Reply:
That is probably correct.